From f149ead6f32662d75039570b2b80e94991e62fdd Mon Sep 17 00:00:00 2001
From: Daniel Kahn Gillmor <dkg@fifthhorseman.net>
Date: Tue, 4 Feb 2025 01:11:08 -0500
Subject: gpg: Reintroduce --openpgp as distinct from --rfc4880

* common/compliance.{c,h}: add CO_OPENPGP
* options.h: #define OPENPGP by analogy with RFC4880
* g10/gpg.c (set_compliance_option): define oOpenPGP separately from
oRFC4880; re-point oDE_VS to use oRFC4880 to avoid changing oDE_VS.
* g10/misc.c (compliance_failure): Describe standards modes
* doc/gpg.texi: (Compliance options): distinguish between --rfc4880
and --openpgp
--

This change adopts baseline MTI digest algorithms for modern OpenPGP
when --compliance=openpgp is supplied.

Signed-off-by: Daniel Kahn Gillmor <dkg@fifthhorseman.net>

Rebased from FreePG source-2.5.21-freepg for GnuPG 2.5.24. Retain the upstream FIPS compliance mode.
---
diff --git a/common/compliance.c b/common/compliance.c
--- a/common/compliance.c
+++ b/common/compliance.c
@@ -85,6 +85,7 @@
 get_compliance_cache (enum gnupg_compliance_mode compliance, int for_rng)
 {
   static int r_gnupg   = -1, s_gnupg   = -1;
+  static int r_openpgp = -1, s_openpgp = -1;
   static int r_rfc4880 = -1, s_rfc4880 = -1;
   static int r_rfc2440 = -1, s_rfc2440 = -1;
   static int r_pgp7    = -1, s_pgp7    = -1;
@@ -97,6 +98,7 @@
   switch (compliance)
     {
     case CO_GNUPG:   ptr = for_rng? &r_gnupg   : &s_gnupg  ; break;
+    case CO_OPENPGP: ptr = for_rng? &r_openpgp : &s_openpgp; break;
     case CO_RFC4880: ptr = for_rng? &r_rfc4880 : &s_rfc4880; break;
     case CO_RFC2440: ptr = for_rng? &r_rfc2440 : &s_rfc2440; break;
     case CO_PGP7:    ptr = for_rng? &r_pgp7    : &s_pgp7   ; break;
@@ -1052,6 +1054,7 @@
     {
     case CO_GNUPG:
       return "8";
+    case CO_OPENPGP:
     case CO_RFC4880:
     case CO_RFC2440:
     case CO_PGP7:
@@ -1184,7 +1187,8 @@
   switch (compliance)
     {
     case CO_GNUPG:   return "--compliance=gnupg";
-    case CO_RFC4880: return "--compliance=openpgp";
+    case CO_OPENPGP: return "--compliance=openpgp";
+    case CO_RFC4880: return "--compliance=rfc4880";
     case CO_RFC2440: return "--compliance=rfc2440";
     case CO_PGP7:    return "--compliance=pgp7";
     case CO_PGP8:    return "--compliance=pgp8";
diff --git a/common/compliance.h b/common/compliance.h
--- a/common/compliance.h
+++ b/common/compliance.h
@@ -38,7 +38,7 @@
 
 enum gnupg_compliance_mode
   {
-    CO_GNUPG, CO_RFC4880, CO_RFC2440,
+    CO_GNUPG, CO_OPENPGP, CO_RFC4880, CO_RFC2440,
     CO_PGP7, CO_PGP8, CO_DE_VS, CO_FIPS
   };
 
diff --git a/doc/gpg.texi b/doc/gpg.texi
--- a/doc/gpg.texi
+++ b/doc/gpg.texi
@@ -3177,18 +3177,15 @@
 
 @item --openpgp
 @opindex openpgp
-Reset all packet, cipher and digest options to strict OpenPGP
-behavior.  This option implies @option{--allow-old-cipher-algos}.  Use
-this option to reset all previous options like @option{--s2k-*},
-@option{--cipher-algo}, @option{--digest-algo} and
-@option{--compress-algo} to OpenPGP compliant values. All PGP
-workarounds are disabled.
+Set all packet, cipher and digest options to OpenPGP compatible
+(RFC-9580) behavior.  Note that not all of RFC-9580 is implemented by
+GnuPG.
 
 @item --rfc4880
 @opindex rfc4880
-Reset all packet, cipher and digest options to strict RFC-4880
-behavior.  This option implies @option{--allow-old-cipher-algos}.
-Note that this is currently the same thing as @option{--openpgp}.
+Set all packet, cipher and digest options to strict RFC-4880 behavior.
+RFC-4880 is the legacy version of the OpenPGP standard.  This option
+implies @option{--allow-old-cipher-algos}.
 
 @item --rfc4880bis
 @opindex rfc4880bis
diff --git a/g10/gpg.c b/g10/gpg.c
--- a/g10/gpg.c
+++ b/g10/gpg.c
@@ -2341,6 +2341,12 @@
       break;
 
     case oOpenPGP:
+      /* use MTIs from RFC 9580 at least: */
+      set_compliance_option (oGnuPG);
+      opt.compliance = CO_OPENPGP;
+      opt.flags.dsa2 = 1;
+      opt.allow_freeform_uid = 1;
+      break;
     case oRFC4880:
       /* This is effectively the same as RFC2440, but with
          "--enable-dsa2 --no-rfc2440-text --escape-from-lines
diff --git a/g10/misc.c b/g10/misc.c
--- a/g10/misc.c
+++ b/g10/misc.c
@@ -1484,12 +1484,16 @@
       ver="GnuPG";
       break;
 
-    case CO_RFC4880:
+    case CO_OPENPGP:
       ver="OpenPGP";
       break;
 
+    case CO_RFC4880:
+      ver="OpenPGP (older, RFC 4880)";
+      break;
+
     case CO_RFC2440:
-      ver="OpenPGP (older)";
+      ver="OpenPGP (oldest, RFC 2440)";
       break;
 
     case CO_PGP7:
diff --git a/g10/options.h b/g10/options.h
--- a/g10/options.h
+++ b/g10/options.h
@@ -418,6 +418,7 @@
 /* Compliance test macros.  */
 #define GNUPG   (opt.compliance==CO_GNUPG || opt.compliance==CO_DE_VS \
                  || opt.compliance==CO_FIPS)
+#define OPENPGP (opt.compliance==CO_OPENPGP)
 #define RFC2440 (opt.compliance==CO_RFC2440)
 #define RFC4880 (opt.compliance==CO_RFC4880)
 #define PGP7    (opt.compliance==CO_PGP7)
